wmbutton/wmbutton.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH WMBUTTON 1 "August 20, 2012"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
wmbutton \- dockable launcher application
.SH SYNOPSIS
.B wmbutton
.\"[-g \fIgeometry\fP] [-d \fIdisplay\fP] [-v] [-f \fIconfigfile\fP] [-h]
[\fIOPTIONS\fP]

.SH DESCRIPTION
This manual page documents briefly the
.B wmbutton
command.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics,
.\" respectively.
\fBwmbutton\fP is a 64x64 pixel application that displays nine buttons and can be used as dockapp for X window managers such as Window Maker, AfterStep, BlackBox, and Enlightenment. Each of these buttons can be configured via a configuration file to run just about any application you would like to. Basically, if you can type it in a shell command, wmbutton can run it.

.PP
\fBwmbutton\fP is configured using a configuration file, ~/.wmbutton (or /etc/wmbutton.conf). Its format is documented in the "Configuration" section below.
.SH OPTIONS
.TP
.B \-g <geometry>
Window Geometry - ie: 64x64+10+10
.TP
.B \-d  <display>
Display -  ie: 127.0.0.1:0.0
.TP
.B \-f  <filename>
Full path to configuration file.
.TP
.B \-b  <filename>
Full path to button xpm.
.TP
.B \-F  <font>
Custom tooltip font (e.g. \-b\&h\-lucidatypewriter\-medium\-*\-*\-*\-12\-*)
.TP
.B \-v
Verbose Mode, gives extra debug output.
.TP
.B \-h
Quick overview of available options.
.TP
.B \-m
Disable Middle Mouse functionality.
.TP
.B \-s
Swap tooltip colors.
.TP
.B \-n
Turn off tooltips.
.TP
.B \-I
Use 80x80 icons.

.SH CONFIGURATION
.TP
The configuration file is case sensitive. A # starts a comment, and lines consisting entirely of whitespace are ignored. The syntax is simply:
.TP
.B <buttonnumber> <command>
.TP
\fBcommand\fP can be anything you can type in as a shell command.
.TP
Buttons are numbered as follows:
.TP
.B Left button
.nf
 	1 2 3
	4 5 6
	7 8 9
.fi
.TP
.B Middle button
.nf
	11 12 13
	14 15 16
	17 18 19
.fi
.TP
.B Right button
.nf
	21 22 23
	24 25 26
	27 28 29
.fi

.TP
To use other icons, wmbutton has to be rebuilt. View /usr/share/doc/wmbutton/README.Debian for further information.

.SH FILES
.TP
.I /etc/wmbutton.conf
system wide wmbutton configuration file.
.TP
.I ~/.wmbutton
per user wmbutton configuration file. If it doesn't exist, the system wide wmbutton configuration file will be used.


.SH AUTHOR
wmbutton was written by Edward H. Flora <ehflora@access1.net>, and is licensed under the terms of the GNU  General Public License.
.TP
This manual page was originaly written by Gordon Fraser <gordon@debian.org>, for the Debian GNU/Linux system (but may be used by others). It was updated by Christian Aichinger <Greek0@gmx.net> and Rodolfo García Peñas (kix) <kix@kix.es>.


ChangeLog
3.0.3 - 8th June 2022.
    - introduced -wallpaper commandline argument to draw live wallpaper. 
 
3.0.2 - 2th June 2022.
    - Fixed compile problem in 2022 environment. 
      https://github.com/norayr/shermans_aquarium/commit/570e5b4137a25d81775b33d1fbed2316a4cccfcd
    - For that gai also had to be changed:
      https://github.com/norayr/gai/commit/c546d6c8161846ae9af2fe6be3299ccec9077181

3.0.1 - 20th February 2005.
    - Fixed compile problem with Linux 2.4 reported by yogib 
      <zban@intermedia.net.pl>
    - Fixed configure problem when you didn't have gai installed, but
      has XMMS installed. (Both gtk 1.2 and 2.0 get included.)
      Reported by Jozef Riha <zefo@seznam.cz> and others.
    - Fixed applet initialization bug noticable sometimes with new gccs.
	
3.0.0 - 5th September 2004. 
    - The current desktop can be used as background in screensaver mode.
    - Sherman's aquarium now interacts with xscreensaver-demo nicely.
    - Fish are now knowing about high aquariums.
    - Bottom creatures shall not come flying.
    - You can have comics on the background in screensaver mode. 
      (some comics included)
    - Fixed bug in plasma with heights >300.
    - Tetris now seems to work..
    - Matrix scroller bug finally fixed.
    - Documentation updates.
    - Thermometer now works with both Linux 2.4 and Linux 2.6
    - Many minor tweeks and fixes.
    - Lots of bugfixes
    - The ChangeLog is incomplete!! (Again!)
    
3.0.0pre2 - 
    - Added patch that solves endian problems for screensaver.
    - Sound effects are quiet when in "over" mode.
    - Screensaver support added, allthough abit buggy when selecting
      from the xscreensaver-demo menu.
    - The screensaver got xscreensaver configuration support.
    - Fixed Gnome panel sizing bug caused by correcting bugs in gai.

3.0.0pre1 - 
    Not complete!
    - Rebirth added.
    - Thermometer pic is now ok.
    - Thermometer can messure a lot of stuff.
    - exploded fishes and eaten fished cooperated now.(Thanks to?)
    - New size organization.
    - Sherman's aquarium is splitted into:
	Sherman's aquarium - Applet
	Sherman's aquarium - Screensaver 
    - Command line options are now completly removed.
      You have to configure Sherman's aquarium via the preference menu.
      It can be started by clicking on it with the right mouse button.
      It works no matter if you use gnome or not.
    - You can have a background shade between any two colours.
    - Great parts are rewritten and the code is modulized.
    - Code "prettification" and reorganization. Still alot to do.
    - Uses now the "General Applet Interface Library" (GAI) for basic
      wmapplet/dockapp and gnome2 applet routines. (Gnome 1.x is no
      more supported.) 
    - Removed some warnings on Redhat systems.(Thanks to Gerald Schuberth!)
    - More plants and rocks.
    - Megan added. 
    - Fixed fvwm2 problem. (Thanks to Fabio Stumbo!)
    - Fixed SDL problem that made Sherman's living on after quit and
      eating alot of processor power.
    - Made fish scared by the mouse pointer acting better. 
    - No outputs in xscreensaver mode.
    - Hiding mouse pointer when over applet option added.
    - Matrix scroller is partly rewritten. (Bug fixed hopefully)
    - Tetris with levels, score and highscore table is added.
    - Added code for a plasma.
    - Now it is possible to load and save settings without compile 
      in gnome support. (Thanks to GAI.)
    - Unnessecary configure options are removed.
    - Added some different bottom colour shades.
    - Bottom now possible option as applet.
    - The scale of the bottom can be set.
    - Fixed bug that made some fish sometimes come behind the moving
      "bottom" creatures.
    - Added more colours of the leds and maximum of four leds instead of
      three. 
    - Fish can have abit variable size, so two fish from the same
      spiece don't have the same size.
    - The speed of each fish variables alittle for the normal speed for
      that spiece.
    - Fixed a rounding up bug that could cause some fish get the images
      of the moment animation somewhat wrong for some scales.
    - Makefile dependency fixes.
    - New feature can display XMMS(www.xmms.org) current song name.
    - Added the original sound effects. Sean Maguire <maguire@mit.edu>
      missed them. 
      
    

2.2.0 - 31st August 2002
    - Makefile & configure fixes.
    - Swedish and German translation for the Gnome menu.
      (But not yet for the preference window.)
    - Fixed preference window and keyleds bug.
    - Gnome 2 panel detection. Notice that gnome 2 applets must be started from
      the panel.
    - XScreenSaver is now supported! Finally!
    - Internal changes makes it possible to run more than one instance at
      the same time. (Screensaver and applet, for example.)
    - Resize blowfish blow up action to correct size. Strangly it was two
      time too big. 
    - The blowfish now explodes and updates correctly in fullscreen.
    - The pray gets eaten correctly now.
    - You can have a seafloor(-f)in fullscreen mode now. 
    - Some changes for SunOs. But still no thermometer code.
    - New graphics taken from the Swedish version of Sherman's Lagoon.
    - Added graphics includes plants and stones for the sea floor.
    - Hawthorne and Bob the bottom dweller are added. They appear only in
      fullscreen/xscreensaver mode.
    - Added new command line option "-m" for activate the matrix scroller
      for those that hasn't gnome installed.

wmbutton/wmbutton.c
/***********************************************************************
 *   Code is based on wmppp, wmload, wmtime, wmcp, and asbutton
 *   Author: Edward H. Flora <ehflora@ksu.edu>
 *
 *   Contributors:
 *              Christian 'Greek0' Aichinger <Greek0@gmx.net>
 *                  Did some code cleanup and fixed several memory leaks.
 *              Ralf Horstmann <ralf.horstmann@gmx.de>
 *                  Added ability to load pixmaps at startup,
 *                  without having to re-compile
 *              Michael Cohrs <camico@users.sourceforge.net>
 *                  Added Tool Tips, and updated graphics
 *              Bruno Essmann <essmann@users.sourceforge.net>)
 *                  Creator of wmpager
 *              Casey Harkins <charkins@cs.wisc.edu>
 *                  Bug fix reading config file path - 3/6/99
 *                  Added button-presses, and other - denoted by *charkins*
 *              Ben Cohen <buddog@aztec.asu.edu>
 *                  original author of wmcp (et al.)
 *              Thomas Nemeth <tnemeth@multimania.com>
 *                  contributor to wmcp
 *              Michael Henderson <mghenderson@lanl.gov>
 *                  Application ideas, suggestions
 *              Ryan ?? <pancake@mindspring.com>
 *                  Modified wmbutton to asbutton.
 *                  Note: asbutton is a seperate program, not associated
 *                        with wmbutton (just as wmbutton is not associated
 *                        with wmcp)
 *              Jon Bruno
 *                  Web Page Development
 *    The contributors listed above are not necessarily involved with the
 *    development of wmbutton.  I'm listing them here partially as thanks for
 *    helping out, catching bugs in the code, etc.
 ***********************************************************************/
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>

#include "wmbutton.h"

#include "backdrop.xpm"           /* background graphic */
#include "buttons.xpm"            /* graphic of 9 buttons */
#include "mask.xbm"               /* Border Graphic */
#include "backdrop-80x80.xpm"	  /* 80x80 icon versions */
#include "buttons-80x80.xpm"
#include "mask-80x80.xbm"



/*************** Function Prototypes ***********************************/
void redraw(void);
void getPixmaps(void);
int  whichButton(int x, int y);   // determine which button has been pressed
void SetWmHints();
void SetClassHints();

/***********************************************************************
 * 		Globals..    OK.. there's too many globals.
 *                           Feel free and fix it, if you'd like.
 ***********************************************************************/
Display *display;
int screen;
Window rootwin, win, iconwin;
GC gc;
int depth;
Pixel bg_pixel, fg_pixel;

struct Config_t Config;

typedef struct _XpmIcon {
	Pixmap pixmap;
	Pixmap mask;
	XpmAttributes attributes;
} XpmIcon;

typedef struct _button_region {
	int x, y;
	int i, j;
} ButtonArea;

ButtonArea button_region[9];

XpmIcon template, visible, buttons;

int border = 0;
int button_pressed = -1;	/* button to be drawn pressed *charkins*/

char *app_name = "wmbutton";

/***********************************************************************
 * 		Main
 ***********************************************************************/
int main(int argc, char **argv)
{
	XEvent report;
	XGCValues xgcValues;
	XTextProperty app_name_atom;
	XSizeHints xsizehints;
	Pixmap pixmask;

	int dummy = 0;
	int N = 1;	/* Button number pressed to goto app # */

	/* Added for Tool Tip Support */
	long nTooltipShowDelay = TOOLTIP_SHOW_DELAY;
	long nTooltipReshowDelay = TOOLTIP_RESHOW_DELAY;
	long nTooltipTimer = -1;
	long nTooltipHideTimer = -1;
	long nNow;
	int nTooltipButton = 0, nTooltipX = 0, nTooltipY = 0;
	int mask_width, mask_height;

	/* Parse Command Line Arguments */
	parseargs(argc, argv);

	if (Config.bigicon == 1) {
	  /* printf( "Big icon\n");*/
	  mask_width = 80;
	  mask_height = 80;
	} else {
	  mask_width = 64;
	  mask_height = 64;
	}



	/* Catch fire if no configuration file exists */
	if (!canOpenFile(Config.configfile)) {
		if(!canOpenFile(CONFIGGLOBAL)) {
			err_mess(FAILCONF, Config.configfile);
			return (1);
		}
	}

	/* Open Display */
	if ((display = XOpenDisplay(Config.Display_str)) == NULL)
		err_mess(FAILDISP, Config.Display_str);

	screen  = DefaultScreen(display);
	rootwin = RootWindow(display, screen);
	depth   = DefaultDepth(display, screen);

	bg_pixel = WhitePixel(display, screen);
	fg_pixel = BlackPixel(display, screen);

	xsizehints.flags  = USSize | USPosition;
	xsizehints.width  = mask_width;
	xsizehints.height = mask_height;

	/* Parse Geometry string and fill in sizehints fields */
	XWMGeometry(display, screen,
		    Config.Geometry_str,
		    NULL,
		    border,
		    &xsizehints,
		    &xsizehints.x,
		    &xsizehints.y,
		    &xsizehints.width,
		    &xsizehints.height,
		    &dummy);

	if ((win = XCreateSimpleWindow(display,
				       rootwin,
				       xsizehints.x,
				       xsizehints.y,
				       xsizehints.width,
				       xsizehints.height,
				       border,
				       fg_pixel, bg_pixel)) == 0)
		err_mess(FAILSWIN, NULL);

	if ((iconwin = XCreateSimpleWindow(display,
					   win,
					   xsizehints.x,
					   xsizehints.y,
					   xsizehints.width,
					   xsizehints.height,
					   border,
					   fg_pixel, bg_pixel)) == 0)

		err_mess(FAILICON, NULL);

	/* Set up shaped windows */
	/* Gives the appicon a border so you can grab and move it. */
	if (Config.bigicon == 0) {
	  /* 64x64 icon path */
	  if ((pixmask = XCreateBitmapFromData(display,
					       win,
					       mask_bits,
					       mask_width,
					       mask_height)) == 0)
	    err_mess(FAILXPM, NULL);
	} else {
	  /* 80x80 icon path */
	  if ((pixmask = XCreateBitmapFromData(display,
					       win,
					       mask_80x80_bits,
					       mask_width,
					       mask_height)) == 0)
	    err_mess(FAILXPM, NULL);
	}
	  
	XShapeCombineMask(display, win, ShapeBounding, 0, 0, pixmask, ShapeSet);
	XShapeCombineMask(display, iconwin, ShapeBounding, 0, 0, pixmask, ShapeSet);

	/* Convert in pixmaps from .xpm includes. */
	getPixmaps();

	/* Interclient Communication stuff */
	/* Appicons don't work with out this stuff */
	SetWmHints();
	SetClassHints();

	XSetWMNormalHints(display, win, &xsizehints);

	/* Tell window manager what the title bar name is. We never see */
	/* this anyways in the WithdrawnState      */
	if (XStringListToTextProperty(&app_name, 1, &app_name_atom) == 0)
		err_mess(FAILWNAM, app_name);

	XSetWMName(display, win, &app_name_atom);

	/* Create Graphic Context */
	if ((gc = XCreateGC(display, win, (GCForeground | GCBackground),
			    &xgcValues)) == NULL)
		err_mess(FAILGC, NULL);

	/* XEvent Masks. We want both window to process X events */
	XSelectInput(display, win,
		     ExposureMask |
		     ButtonPressMask |
		     ButtonReleaseMask | /* added ButtonReleaseMask *charkins*/
		     PointerMotionMask |
		     StructureNotifyMask |
		     LeaveWindowMask);

	XSelectInput(display, iconwin,
		     ExposureMask |
		     ButtonPressMask |
		     ButtonReleaseMask | /* added ButtonReleaseMask *charkins*/
		     PointerMotionMask |
		     StructureNotifyMask |
		     LeaveWindowMask);

	/* Store the 'state' of the application for restarting */
	XSetCommand(display, win, argv, argc);

	/* Window won't ever show up until it is mapped.. then drawn after a 	*/
	/* ConfigureNotify */
	XMapWindow(display, win);

	/* Initialize Tooltip Support */
	initTime();
	initTooltip();

	/* X Event Loop */
	while (1) {
		while (XPending(display) || nTooltipTimer == -1) {
			XNextEvent(display, &report);

			switch (report.type) {
			case Expose:
				if (report.xexpose.count != 0)
					break;

				if (Config.Verbose)
					fprintf(stdout, "Event: Expose\n");

				redraw();
				break;

			case ConfigureNotify:
				if (Config.Verbose)
					fprintf(stdout, "Event: ConfigureNotify\n");

				redraw();
				break;

			case MotionNotify:
				if (hasTooltipSupport()) {
					if (!hasTooltip()) {
						nTooltipTimer = currentTimeMillis();
						nTooltipX = report.xbutton.x;
						nTooltipY = report.xbutton.y;
						nTooltipButton = whichButton(report.xbutton.x, report.xbutton.y);
					} else {
						int nButton = whichButton(report.xbutton.x, report.xbutton.y);
						if (nButton != nTooltipButton) {
							hideTooltip();
							nTooltipTimer = -1;
							nTooltipX = report.xbutton.x;
							nTooltipY = report.xbutton.y;
							nTooltipButton = nButton;
							showTooltip(nTooltipButton, nTooltipX, nTooltipY);
						}
					}
				}
				break;

			case LeaveNotify:
				if (Config.Verbose)
					fprintf(stdout, "Event: LeaveNotify\n");

				if (hasTooltip()) {
					hideTooltip();
					nTooltipHideTimer = currentTimeMillis();
				}
				nTooltipTimer = -1;
				break;

			case ButtonPress:	/* draw button pressed, don't launch *charkins*/
				if (hasTooltip()) {
					hideTooltip();
					nTooltipHideTimer = currentTimeMillis();
				}

				switch (report.xbutton.button) {
				case Button1:
					N = whichButton(report.xbutton.x, report.xbutton.y);
					if ((N >= 0) && (N <= NUMB_OF_APPS)) {
						button_pressed = N + LMASK;
						redraw();
					}

					if (Config.Verbose)
						fprintf(stdout, "Button 1:x=%d y=%d N=%d\n",
							report.xbutton.x, report.xbutton.y, N+LMASK);
					break;

				case Button2:
					if (!Config.mmouse) {
						N = whichButton(report.xbutton.x, report.xbutton.y);
						if ((N >= 0) && (N <= NUMB_OF_APPS)) {
							button_pressed = N + MMASK;
							redraw();
						}

						if (Config.Verbose)
							fprintf(stdout, "Button 2:x=%d y=%d N=%d\n",
								report.xbutton.x, report.xbutton.y, N+MMASK);
					}
					break;

				case Button3:
					N = whichButton(report.xbutton.x, report.xbutton.y);
					if ((N >= 0) && (N <= NUMB_OF_APPS)) {
						button_pressed = N + RMASK;
						redraw();
					}

					if (Config.Verbose)
						fprintf(stdout, "Button 3:x=%d y=%d N=%d\n",
							report.xbutton.x, report.xbutton.y, N+RMASK);
					break;
				}
				break;

			case ButtonRelease:   /* launch app here if still over button *charkins*/
				switch (report.xbutton.button) {
				case Button1:
					N = whichButton(report.xbutton.x, report.xbutton.y);
					if ((N >= 0) && (N <= NUMB_OF_APPS) && (N == button_pressed))
						RunAppN(N + LMASK);

					button_pressed = -1;
					redraw();

					if (Config.Verbose)
						fprintf(stdout, "Button 1:x=%d y=%d N=%d\n",
							report.xbutton.x, report.xbutton.y, N+LMASK);
					break;

				case Button2:
					if (!Config.mmouse) {
						N = whichButton(report.xbutton.x, report.xbutton.y);
						if ((N >= 0) && (N <= NUMB_OF_APPS) && (N == button_pressed))
							RunAppN(N + MMASK);

						button_pressed = -1;
						redraw();

						if (Config.Verbose)
							fprintf(stdout, "Button 2:x=%d y=%d N=%d\n",
								report.xbutton.x, report.xbutton.y, N+MMASK);
					}
					break;

				case Button3:
					N = whichButton(report.xbutton.x, report.xbutton.y);
					if ((N >= 0) && (N <= NUMB_OF_APPS) && (N == button_pressed))
						RunAppN(N + RMASK);

					button_pressed = -1;
					redraw();

					if (Config.Verbose)
						fprintf(stdout, "Button 3:x=%d y=%d N=%d\n",
							report.xbutton.x, report.xbutton.y, N+RMASK);
					break;
				}
				break;

			case DestroyNotify:
				if (Config.Verbose)
					fprintf(stdout, "Bye\n");

				destroyTooltip();
				XFreeGC(display, gc);
				XDestroyWindow(display, win);
				XDestroyWindow(display, iconwin);
				XCloseDisplay(display);
				exit(0);
				break;
			}
		}

		usleep(50000);
		nNow = currentTimeMillis();
		if (nTooltipTimer != -1 &&
		    ((nNow > nTooltipTimer + nTooltipShowDelay) ||
		     (nNow < nTooltipHideTimer + nTooltipReshowDelay))) {
			showTooltip(nTooltipButton, nTooltipX, nTooltipY);
			nTooltipTimer = -1;
		}
	}

	return (0);
}
/***********************************************************************/

/***********************************************************************
 *   redraw
 *
 *	 Map the button region coordinates.
 *
 *   Draw the appropriate number of buttons on the 'visible' Pixmap
 *   using data from the 'buttons' pixmap.
 *
 *   Then, copy the 'visible' pixmap to the two windows ( the withdrawn
 *   main window and the icon window which is the main window's icon image.)
 ***********************************************************************/
void redraw() {
  int n, i, j, dest_x, dest_y, space, offset, bsize; 

	if (Config.Verbose)
		fprintf(stdout, "In Redraw()\n");
	if (Config.bigicon != 1) { /* 18 for 54pixel grid, 22 for 66pixel grid 80x80 */
	  bsize = 18;
	} else {
	  bsize = 22;
	}

	space = 0;
	offset = 5;
	XCopyArea(display, template.pixmap, visible.pixmap, gc, 0, 0,
		  template.attributes.width, template.attributes.height, 0, 0);

	for (j = 0; j < 3; j++) {
		for (i = 0; i < 3; i++) {
			n = i + j * 3;
			dest_x = i * (bsize + space) + offset + space;
			dest_y = j * (bsize + space) + offset + space;

			/* Define button mouse coords */
			button_region[n].x = dest_x;
			button_region[n].y = dest_y;
			button_region[n].i = dest_x + bsize - 1;
			button_region[n].j = dest_y + bsize - 1;

			/* Copy button images for valid apps */
			if ((n + 1) <= NUMB_OF_APPS)
				XCopyArea(display, buttons.pixmap, visible.pixmap, gc,
					  i * bsize, j * bsize, bsize, bsize, dest_x, dest_y);
		}
	}

	if (button_pressed > 0) {	/* draw pressed button *charkins*/
		if (button_pressed > RMASK)
			button_pressed -= RMASK;
		else if (button_pressed > MMASK)
			button_pressed -= MMASK;
		else if (button_pressed > LMASK)
			button_pressed -= LMASK;

		i = (button_pressed - 1) % 3;	/* get col of button */
		j = (button_pressed - 1) / 3;	/* get row of button */
		dest_x = i * (bsize + space) + offset + space;
		dest_y = j * (bsize + space) + offset + space;
		XSetForeground(display, gc, bg_pixel);
		XDrawLine(display, visible.pixmap, gc,
			  dest_x + 1, dest_y + bsize - 1,
			  dest_x + bsize - 1, dest_y + bsize - 1);
		XDrawLine(display, visible.pixmap, gc,
			  dest_x + bsize - 1, dest_y + bsize - 1,
			  dest_x + bsize - 1, dest_y + 1);
		XSetForeground(display, gc, fg_pixel);
		XDrawLine(display, visible.pixmap, gc,
			  dest_x, dest_y, dest_x + bsize - 2, dest_y);
		XDrawLine(display, visible.pixmap, gc,
			  dest_x, dest_y, dest_x, dest_y + bsize - 2);
	} /*charkins*/

	flush_expose(win);
	XCopyArea(display, visible.pixmap, win, gc, 0, 0,
		  visible.attributes.width, visible.attributes.height, 0, 0);
	flush_expose(iconwin);
	XCopyArea(display, visible.pixmap, iconwin, gc, 0, 0,
		  visible.attributes.width, visible.attributes.height, 0, 0);
}
/***********************************************************************/

/***********************************************************************
 *  whichButton
 *
 *  Return the button that at the x,y coordinates. The button need not
 *  be visible ( drawn ). Return -1 if no button match.
 ***********************************************************************/
int whichButton(int x, int y)
{
	int index;

	for (index = 0; index < NUMB_OF_APPS; index++) {
		if (x >= button_region[index].x &&
		    x <= button_region[index].i &&
		    y >= button_region[index].y &&
		    y <= button_region[index].j)
			return(index + 1);
	}
	return -1;
}
/***********************************************************************/


/***********************************************************************
 *   getPixmaps
 *
 *   Load XPM data into X Pixmaps.
 *
 *   Pixmap template contains the untouched window backdrop image.
 *   Pixmap visible is the template pixmap with buttons drawn on it.
 *          -- what is seen by the user.
 *   Pixmap buttons holds the images for individual buttons that are
 *          later copied onto Pixmap visible.
 ***********************************************************************/
void getPixmaps()
{
	int loaded = 0;
	template.attributes.valuemask |= (XpmReturnPixels | XpmReturnExtensions);
	visible.attributes.valuemask  |= (XpmReturnPixels | XpmReturnExtensions);
	buttons.attributes.valuemask  |= (XpmReturnPixels | XpmReturnExtensions);

	if (Config.Verbose)
		fprintf(stdout, "In getPixmaps\n");

	if (Config.bigicon == 0) {  
	  /* 64x64 icon path.  Template Pixmap. Never Drawn To. */
	  if (XpmCreatePixmapFromData(display, rootwin, backdrop_xpm,
				&template.pixmap, &template.mask,
				&template.attributes) != XpmSuccess)
		err_mess(FAILTMPL, NULL);

	/* Visible Pixmap. Copied from template Pixmap and then drawn to. */
	  if (XpmCreatePixmapFromData(display, rootwin, backdrop_xpm,
				&visible.pixmap, &visible.mask,
				&visible.attributes) != XpmSuccess)
		err_mess(FAILVIS, NULL);

	/* Button Pixmap.  */
	  if (access(Config.buttonfile, R_OK) == 0) {
		/* load buttons from file */
	    if (XpmReadFileToPixmap(display, rootwin, Config.buttonfile,
					&buttons.pixmap, &buttons.mask,
					&buttons.attributes) != XpmSuccess)
			err_mess(FAILBUT, NULL);
	    else
			loaded = 1;
	  }

	  if (!loaded) {
		/* Use Builtin Button Pixmap.  */
	    if (Config.Verbose)
	      fprintf(stdout, "Using builtin buttons pixmap 64x64\n");

	    if (XpmCreatePixmapFromData(display, rootwin, buttons_xpm,
					&buttons.pixmap, &buttons.mask,
					&buttons.attributes) != XpmSuccess)
			err_mess(FAILBUT, NULL);
	  }

	  if (Config.Verbose)
		fprintf(stdout, "Leaving small icon getPixmaps\n");
	} else { 
	  /* 80x80 icon path. Template Pixmap. Never Drawn To. */
	  if (XpmCreatePixmapFromData(display, rootwin, backdrop_80x80_xpm,
				&template.pixmap, &template.mask,
				&template.attributes) != XpmSuccess)
		err_mess(FAILTMPL, NULL);

	/* Visible Pixmap. Copied from template Pixmap and then drawn to. */
	  if (XpmCreatePixmapFromData(display, rootwin, backdrop_80x80_xpm,
				&visible.pixmap, &visible.mask,
				&visible.attributes) != XpmSuccess)
		err_mess(FAILVIS, NULL);

	/* Button Pixmap.  */
	  if (access(Config.buttonfile, R_OK) == 0) {
		/* load buttons from file */
	    if (XpmReadFileToPixmap(display, rootwin, Config.buttonfile,
					&buttons.pixmap, &buttons.mask,
					&buttons.attributes) != XpmSuccess)
			err_mess(FAILBUT, NULL);
	    else
			loaded = 1;
	  }

	  if (!loaded) {
		/* Use Builtin Button Pixmap.  */
	    if (Config.Verbose)
	      fprintf(stdout, "Using builtin buttons pixmap 80x80\n");

	    if (XpmCreatePixmapFromData(display, rootwin, buttons_80x80_xpm,
					&buttons.pixmap, &buttons.mask,
					&buttons.attributes) != XpmSuccess)
			err_mess(FAILBUT, NULL);
	  }

	  if (Config.Verbose)
		fprintf(stdout, "Leaving big icon getPixmaps\n");
	}	  
}
/*********************************************************************/

void SetWmHints()
{
	XWMHints *xwmhints;

	xwmhints = XAllocWMHints();
	xwmhints->flags = WindowGroupHint | IconWindowHint | StateHint;
	xwmhints->icon_window = iconwin;
	xwmhints->window_group = win;
	xwmhints->initial_state = WithdrawnState;
	XSetWMHints(display, win, xwmhints);
	XFree(xwmhints);
	xwmhints = NULL;
}

void SetClassHints()
{
	XClassHint xclasshint;

	xclasshint.res_name = "wmbutton";
	xclasshint.res_class = "Wmbutton";
	XSetClassHint(display, win, &xclasshint);
}
